test/util/framework/per_test_framework.go

```diff
@@ -0,0 +1,273 @@
+	} else {
+		tc.BeforeEach(context.TODO())
+	}
+
+	return tc
+}
+
+// BeforeEach gives a chance for initialization (none yet) and registers the cleanup
+func (tc *perItOrDescribeTestContext) BeforeEach(ctx context.Context) {
+	// DeferCleanup, in contrast to AfterEach, triggers execution in
+	// first-in-last-out order. This ensures that the framework instance
+	// remains valid as long as possible.
+	//
+	// In addition, AfterEach will not be called if a test never gets here.
+	ginkgo.DeferCleanup(tc.deleteCreatedResources, AnnotatedLocation("tear down test context"), ginkgo.NodeTimeout(45*time.Minute))
+
+	// Registered later and thus runs before deleting namespaces.
+	ginkgo.DeferCleanup(tc.collectDebugInfo, AnnotatedLocation("dump debug info"), ginkgo.NodeTimeout(45*time.Minute))
+}
+
+// deleteCreatedResources deletes what was created that we know of.
+func (tc *perItOrDescribeTestContext) deleteCreatedResources(ctx context.Context) {
+	tc.contextLock.RLock()
+	defer tc.contextLock.RUnlock()
+
+	// deletion takes a while, it's worth it to do this in parallel
+	waitGroup, ctx := errgroup.WithContext(ctx)
+	for _, resourceGroupName := range tc.knownResourceGroups {
+		waitGroup.Go(func() error {
+			// prevent a stray panic from exiting the process. Don't do this generally because ginkgo/gomega rely on panics ot funcion.
+			utilruntime.HandleCrashWithContext(ctx)
+
+			return tc.cleanupResourceGroup(ctx, resourceGroupName)
+		})
+	}
+	if err := waitGroup.Wait(); err != nil {
+		// remember that Wait only shows the first error, not all the errors.
+		ginkgo.GinkgoLogr.Error(err, "at least one resource group failed to delete: %w", err)
+	}
+}
+
+// collectDebugInfo collects information and saves it in artifact dir
+func (tc *perItOrDescribeTestContext) collectDebugInfo(ctx context.Context) {
+	// TODO dump deployment operation info for all resource groups
+}
+
+func (tc *perItOrDescribeTestContext) NewResourceGroup(ctx context.Context, resourceGroupPrefix, location string) (*armresources.ResourceGroup, error) {
+
+	suffix := rand.String(6)
+	resourceGroupName := SuffixName(resourceGroupPrefix, suffix, 64)
+	func() {
+		tc.contextLock.Lock()
+		defer tc.contextLock.Unlock()
+		tc.knownResourceGroups = append(tc.knownResourceGroups, resourceGroupName)
+	}()
+	ginkgo.GinkgoLogr.Info("creating resource group", "resourceGroup", resourceGroupName)
+
+	resourceGroup, err := CreateResourceGroup(ctx, tc.GetARMResourcesClientFactoryOrDie(ctx).NewResourceGroupsClient(), resourceGroupName, location, 20*time.Minute)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create resource group: %w", err)
+	}
+
+	return resourceGroup, nil
+}
+
+// cleanupResourceGroup is the standard resourcegroup cleanup.  It attempts to
+// 1. delete all HCP clusters and wait for success
+// 2. delete the resource group and wait for success
+func (tc *perItOrDescribeTestContext) cleanupResourceGroup(ctx context.Context, resourceGroupName string) error {
```

Comment:
Does this take a slot in your ginkgo parallelism? I imagine RG cleanup may take a long time but not cost any resources client-side